File: javascript/nodes/save_m3u_file.js
import { Next } from '../ambler.js';
import fs from 'fs';

const defaultUtils = { writeTextFile: (path, content) => fs.writeFileSync(path, content) };

export function saveM3UFile(edges, utils = defaultUtils) {
    return async (state) => {
        if (!state.m3uFilePath) {
            throw new Error('M3U file path is not defined.');
        }

        const content = state.urls.join('\n');
        utils.writeTextFile(state.m3uFilePath, content);
        console.log(`Saved resolved URLs to ${state.m3uFilePath}`);
        return new Next(edges.onSuccess, state);
    };
}

File: python/nodes/save_m3u_file.py
from typing import Callable, TypedDict

from ambler import Next, Nextable
from state import State


class SaveEdges(TypedDict):
    on_success: Nextable


class SaveUtils(TypedDict):
    write_text_file: Callable[[str, str], None]


def _write_text_file(path: str, content: str) -> None:
    with open(path, 'w') as f:
        f.write(content)


_default_utils: SaveUtils = {
    'write_text_file': _write_text_file,
}


def save_m3u_file(edges: SaveEdges, utils: SaveUtils = _default_utils) -> Nextable:
    async def _node(state: State) -> Next[State]:
        if not state['m3u_file_path']:
            raise ValueError("M3U file path is not defined.")

        content = '\n'.join(state['urls'])
        utils['write_text_file'](state['m3u_file_path'], content)
        print(f"Saved resolved URLs to {state['m3u_file_path']}")
        return Next(edges['on_success'], state)

    return _node
